src/twisted/web/http.py

```diff
@@ -377,6 +379,75 @@ def parseContentRange(header):
 
 
 
+class _IDeprecatedHTTPChannelToRequestInterface(Interface):
+    """
+    The interface L{HTTPChannel} expects of L{Request}.
+    """
+
+    requestHeaders = Attribute(
+        "A L{http_headers.Headers} instance giving all received HTTP request "
+        "headers.")
+
+    responseHeaders = Attribute(
+        "A L{http_headers.Headers} instance holding all HTTP response "
+        "headers to be sent.")
+
+
+    def connectionLost(reason):
+        """
+        The underlying connection has been lost.
+
+        @param reason: A failure instance indicating the reason why
+            the connection was lost.
+        @type reason: L{twisted.python.failure.Failure}
+        """
+
+
+    def gotLength(length):
+        """
+        Called when L{HTTPChannel} has determined the length, if any,
+        of the incoming request's body.
+
+        @param length: The length of the request's body.
+        @type length: L{int} if the request declares its body's length
+            and L{None} if it does not.
+        """
+
+
+    def handleContentChunk(data):
+        """
+        Deliver a received chunk of body data to the request.  Note
+        this does not imply chunked transfer encoding.
+
+        @param data: The received chunk.
+        @type data: L{bytes}
+        """
+
+
+    def parseCookies():
+        """
+        Parse the request's cookies out of received headers.
+        """
+
+
+    def requestReceived(command, path, version):
+        """
+        Called when the entire request, including its body, has been
+        received.
+
+        @param command: The request's HTTP command.
+        @type command: L{bytes}
+
+        @param path: The request's path.  Note: this is actually what
+            RFC7320 calls the URI.
+        @type path: L{bytes}
+
+        @param version: The request's HTTP version.
+        @type version: L{bytes}
+        """
+
+
```

Comment:
OK, so the trouble with adding `__eq__`/`__ne__` to the interface and implementing `Request.__eq__`/`__ne__` is that the proxied class gets compared to a _wrapped_ instance.

Specifically:

```python
if request != self.requests[0]: raise TypeError
```

ends up being 

```python
if requests.__eq__(self.requests[0]): raise TypeError
```

the guts of this end up being

```python
def __eq__(self, other):
    return self is other
```

since the `__eq__` implementation gets proxied through to the underlying request object, that in turns ends up being

```
return actualRequestInstance is not proxyAroundActualRequestInstance
```

...which is of course `False`.  i'm going to claim this difficulty is exactly why i did what i did with the tests' proxy subclass.

i could add an identity function to the interface that returned `id(self)`, but that seems excessive.  is there an easy to do this i'm not seeing?